File: python/server/manifest_server.py
"""TCP server for fixed-block manifest requests."""

from __future__ import annotations

import hashlib
import hmac
import json
import os
import socket
import struct
import threading
import zlib
from pathlib import Path

from scripts.fixed_block_analysis import (
    DEFAULT_BLOCK_SIZE,
    build_manifest,
    recv_message,
    send_message,
)


def _send_error(client_socket: socket.socket, code: str) -> None:
    send_message(client_socket, json.dumps({"error": code}).encode("utf-8"))


def _resolve_file_path(base_dir: Path, filename: str) -> Path | None:
    file_path = (base_dir / filename).resolve()
    if base_dir not in file_path.parents and file_path != base_dir:
        return None
    return file_path


def _compute_file_hash(path: Path, block_size: int) -> str:
    digest = hashlib.sha256()
    with path.open("rb") as handle:
        while True:
            chunk = handle.read(block_size)
            if not chunk:
                break
            digest.update(chunk)
    return digest.hexdigest()


def _handle_get_manifest(
    client_socket: socket.socket,
    base_dir: Path,
    block_size: int,
    payload: dict[str, object],
) -> None:
    filename = payload.get("filename")
    if not isinstance(filename, str):
        _send_error(client_socket, "INVALID_FILENAME")
        return

    file_path = _resolve_file_path(base_dir, filename)
    if file_path is None:
        _send_error(client_socket, "INVALID_PATH")
        return

    if not file_path.exists() or not file_path.is_file():
        _send_error(client_socket, "NOT_FOUND")
        return

    manifest = build_manifest(file_path, block_size=block_size)
    send_message(client_socket, json.dumps({"manifest": manifest}).encode("utf-8"))


def _apply_delta(
    client_socket: socket.socket,
    base_dir: Path,
    payload: dict[str, object],
    default_block_size: int,
    psk: str | None,
) -> None:
    filename = payload.get("filename")
    if not isinstance(filename, str):
        _send_error(client_socket, "INVALID_FILENAME")
        return

    block_size = payload.get("block_size", default_block_size)
    if not isinstance(block_size, int) or block_size <= 0:
        _send_error(client_socket, "INVALID_BLOCK_SIZE")
        return

    file_size = payload.get("file_size")
    if not isinstance(file_size, int) or file_size < 0:
        _send_error(client_socket, "INVALID_FILE_SIZE")
        return

    blocks_count = payload.get("blocks_count")
    if not isinstance(blocks_count, int) or blocks_count < 0:
        _send_error(client_socket, "INVALID_BLOCK_COUNT")
        return

    full_hash = payload.get("full_hash")
    if not isinstance(full_hash, str):
        _send_error(client_socket, "INVALID_FILE_HASH")
        return

    nonce = payload.get("nonce")
    if not isinstance(nonce, str):
        _send_error(client_socket, "INVALID_NONCE")
        return

    signature = payload.get("hmac")
    if not isinstance(signature, str):
        _send_error(client_socket, "INVALID_HMAC")
        return

    if not psk:
        _send_error(client_socket, "PSK_REQUIRED")
        return

    signing_payload = f"{nonce}:{filename}:{file_size}:{block_size}:{full_hash}".encode(
        "utf-8"
    )
    expected = hmac.new(
        psk.encode("utf-8"), signing_payload, hashlib.sha256
    ).hexdigest()
    if not hmac.compare_digest(expected, signature):
        _send_error(client_socket, "HMAC_MISMATCH")
        return

    file_path = _resolve_file_path(base_dir, filename)
    if file_path is None:
        _send_error(client_socket, "INVALID_PATH")
        return

    fd = os.open(file_path, os.O_RDWR | os.O_CREAT)
    with os.fdopen(fd, "r+b") as handle:
        handle.truncate(file_size)
        for _ in range(blocks_count):
            block_message = recv_message(client_socket)
            if not block_message:
                _send_error(client_socket, "EMPTY_BLOCK")
                return
            if len(block_message) < 8:
                _send_error(client_socket, "INVALID_BLOCK_HEADER")
                return
            index, data_len = struct.unpack("!II", block_message[:8])
            compressed = block_message[8:]
            if data_len != len(compressed):
                _send_error(client_socket, "INVALID_BLOCK_LENGTH")
                return
            try:
                data = zlib.decompress(compressed)
            except zlib.error:
                _send_error(client_socket, "DECOMPRESSION_FAILED")
                return
            if len(data) > block_size:
                _send_error(client_socket, "INVALID_BLOCK_SIZE")
                return
            if index * block_size + len(data) > file_size:
                _send_error(client_socket, "INVALID_BLOCK_RANGE")
                return
            handle.seek(index * block_size)
            handle.write(data)

    computed_hash = _compute_file_hash(file_path, block_size)
    if not hmac.compare_digest(computed_hash, full_hash):
        _send_error(client_socket, "HASH_MISMATCH")
        return

    send_message(
        client_socket,
        json.dumps({"status": "OK", "applied_blocks": blocks_count}).encode("utf-8"),
    )


def _handle_client(
    client_socket: socket.socket, base_dir: Path, block_size: int, psk: str | None
) -> None:
    with client_socket:
        try:
            raw = recv_message(client_socket)
            if not raw:
                return
            try:
                payload = json.loads(raw.decode("utf-8"))
            except json.JSONDecodeError:
                _send_error(client_socket, "INVALID_JSON")
                return

            if not isinstance(payload, dict):
                _send_error(client_socket, "INVALID_PAYLOAD")
                return

            command = payload.get("command")
            if command == "GET_MANIFEST":
                _handle_get_manifest(client_socket, base_dir, block_size, payload)
                return
            if command == "APPLY_DELTA":
                _apply_delta(client_socket, base_dir, payload, block_size, psk)
                return

            _send_error(client_socket, "UNKNOWN_COMMAND")
        except ConnectionError:
            return


def start_server(
    host: str,
    port: int,
    base_dir: str | Path,
    block_size: int = DEFAULT_BLOCK_SIZE,
    psk: str | None = None,
) -> None:
    base_path = Path(base_dir).resolve()
    with socket.socket(socket.AF_INET, socket.SOCK_STREAM) as server_socket:
        server_socket.setsockopt(socket.SOL_SOCKET, socket.SO_REUSEADDR, 1)
        server_socket.bind((host, port))
        server_socket.listen()
        print(f"Manifest server listening on {host}:{port}, base dir: {base_path}")
        while True:
            client_socket, _ = server_socket.accept()
            thread = threading.Thread(
                target=_handle_client,
                args=(client_socket, base_path, block_size, psk),
                daemon=True,
            )
            print(f"Accepted connection from {client_socket.getpeername()}")
            thread.start()


if __name__ == "__main__":
    import argparse

    parser = argparse.ArgumentParser(description="Fixed-block manifest server")
    parser.add_argument("host", help="Host to bind")
    parser.add_argument("port", type=int, help="Port to bind")
    parser.add_argument(
        "--base-dir",
        default=".",
        help="Base directory for file lookups",
    )
    parser.add_argument(
        "-b",
        "--block-size",
        type=int,
        default=DEFAULT_BLOCK_SIZE,
        help="Block size in bytes (default: 65536)",
    )
    parser.add_argument(
        "--psk",
        default=os.environ.get("DELTA_SYNC_PSK"),
        help="Pre-shared key for the security handshake (or set DELTA_SYNC_PSK)",
    )

    args = parser.parse_args()
    start_server(
        host=args.host,
        port=args.port,
        base_dir=args.base_dir,
        block_size=args.block_size,
        psk=args.psk,
    )

File: java/src/main/java/com/medai/deltasync/ManifestServer.java
package com.medai.deltasync;

import java.io.IOException;
import java.io.RandomAccessFile;
import java.net.InetSocketAddress;
import java.net.ServerSocket;
import java.net.Socket;
import java.nio.ByteBuffer;
import java.nio.ByteOrder;
import java.nio.charset.StandardCharsets;
import java.nio.file.Files;
import java.nio.file.Path;
import java.util.List;
import java.util.concurrent.ExecutorService;
import java.util.concurrent.Executors;

import org.json.JSONArray;
import org.json.JSONException;
import org.json.JSONObject;

/**
 * Java port of {@code server/manifest_server.py}.
 *
 * <p>Accepts TCP connections and handles two commands:
 * <ul>
 *   <li>{@code GET_MANIFEST} — returns the SHA-256 manifest of a file</li>
 *   <li>{@code APPLY_DELTA}  — receives changed blocks, patches the file in
 *       place, then re-hashes the full file to verify integrity</li>
 * </ul>
 *
 * <p>Path traversal is blocked: every requested filename is resolved relative
 * to {@code baseDir} and rejected if it escapes that directory.
 */
public final class ManifestServer {

    private ManifestServer() { }

    public static void start(String host, int port, Path baseDir, int blockSize, String psk) throws IOException {
        Path basePath = realOrAbsolute(baseDir);

        ExecutorService pool = Executors.newCachedThreadPool(r -> {
            Thread t = new Thread(r, "deltasync-worker");
            t.setDaemon(true);
            return t;
        });

        try (ServerSocket serverSocket = new ServerSocket()) {
            serverSocket.setReuseAddress(true);
            serverSocket.bind(new InetSocketAddress(host, port));
            System.out.println("Manifest server listening on " + host + ":" + port
                    + ", base dir: " + basePath);

            while (!Thread.currentThread().isInterrupted()) {
                Socket client = serverSocket.accept();
                System.out.println("Accepted connection from " + client.getRemoteSocketAddress());
                pool.submit(() -> handleClient(client, basePath, blockSize, psk));
            }
        }
    }

    /**
     * Resolve {@code base} to an absolute, normalized path. Prefers
     * {@code toRealPath()} (which also resolves symlinks, like Python's
     * {@code Path.resolve()}) but falls back to a normalized absolute path
     * if the directory doesn't exist yet.
     */
    private static Path realOrAbsolute(Path base) {
        try {
            return base.toRealPath();
        } catch (IOException e) {
            return base.toAbsolutePath().normalize();
        }
    }

    private static void handleClient(Socket client, Path baseDir, int blockSize, String psk) {
        try (Socket c = client) {
            byte[] raw;
            try {
                raw = Protocol.recvMessage(c);
            } catch (IOException eof) {
                return;     // peer closed before sending anything
            }
            if (raw.length == 0) return;

            JSONObject payload;
            try {
                payload = new JSONObject(new String(raw, StandardCharsets.UTF_8));
            } catch (JSONException e) {
                sendError(c, "INVALID_JSON");
                return;
            }

            Object cmd = payload.opt("command");
            if (!(cmd instanceof String command)) {
                sendError(c, "UNKNOWN_COMMAND");
                return;
            }

            switch (command) {
                case "GET_MANIFEST" -> handleGetManifest(c, baseDir, blockSize, payload);
                case "APPLY_DELTA"  -> applyDelta(c, baseDir, payload, blockSize, psk);
                default             -> sendError(c, "UNKNOWN_COMMAND");
            }
        } catch (IOException ignored) {
            // connection torn down; nothing else we can do
        }
    }

    private static void handleGetManifest(Socket sock, Path baseDir, int blockSize, JSONObject payload)
            throws IOException {
        Object filenameObj = payload.opt("filename");
        if (!(filenameObj instanceof String filename)) {
            sendError(sock, "INVALID_FILENAME");
            return;
        }
        Path filePath = resolveFilePath(baseDir, filename);
        if (filePath == null) {
            sendError(sock, "INVALID_PATH");
            return;
        }
        if (!Files.exists(filePath) || !Files.isRegularFile(filePath)) {
            sendError(sock, "NOT_FOUND");
            return;
        }

        List<ManifestEntry> manifest = BlockAnalysis.buildManifest(filePath, blockSize);
        JSONArray arr = new JSONArray();
        for (ManifestEntry e : manifest) {
            arr.put(new JSONObject()
                    .put("index",      e.index())
                    .put("hash",       e.hash())
                    .put("start_byte", e.startByte()));
        }
        Protocol.sendMessage(sock,
                new JSONObject().put("manifest", arr).toString().getBytes(StandardCharsets.UTF_8));
    }

    private static void applyDelta(Socket sock,
                                   Path baseDir,
                                   JSONObject payload,
                                   int defaultBlockSize,
                                   String psk) throws IOException {
        Object filenameObj = payload.opt("filename");
        if (!(filenameObj instanceof String filename)) {
            sendError(sock, "INVALID_FILENAME");
            return;
        }

        // block_size: must be a positive int
        Object blockSizeObj = payload.opt("block_size");
        int blockSize = defaultBlockSize;
        if (blockSizeObj instanceof Number bn) {
            blockSize = bn.intValue();
        } else if (blockSizeObj != null) {
            sendError(sock, "INVALID_BLOCK_SIZE");
            return;
        }
        if (blockSize <= 0) {
            sendError(sock, "INVALID_BLOCK_SIZE");
            return;
        }

        // file_size: must be a non-negative long
        Object fileSizeObj = payload.opt("file_size");
        if (!(fileSizeObj instanceof Number fsn)) {
            sendError(sock, "INVALID_FILE_SIZE");
            return;
        }
        long fileSize = fsn.longValue();
        if (fileSize < 0) {
            sendError(sock, "INVALID_FILE_SIZE");
            return;
        }

        // blocks_count: must be a non-negative int
        Object blocksCountObj = payload.opt("blocks_count");
        if (!(blocksCountObj instanceof Number bcn)) {
            sendError(sock, "INVALID_BLOCK_COUNT");
            return;
        }
        int blocksCount = bcn.intValue();
        if (blocksCount < 0) {
            sendError(sock, "INVALID_BLOCK_COUNT");
            return;
        }

        if (!(payload.opt("full_hash") instanceof String fullHash)) {
            sendError(sock, "INVALID_FILE_HASH");
            return;
        }
        if (!(payload.opt("nonce") instanceof String nonce)) {
            sendError(sock, "INVALID_NONCE");
            return;
        }
        if (!(payload.opt("hmac") instanceof String signature)) {
            sendError(sock, "INVALID_HMAC");
            return;
        }

        if (psk == null || psk.isEmpty()) {
            sendError(sock, "PSK_REQUIRED");
            return;
        }

        String signingPayload = nonce + ":" + filename + ":" + fileSize + ":"
                              + blockSize + ":" + fullHash;
        String expected = CryptoUtil.hmacSha256Hex(psk, signingPayload);
        if (!CryptoUtil.constantTimeEquals(expected, signature)) {
            sendError(sock, "HMAC_MISMATCH");
            return;
        }

        Path filePath = resolveFilePath(baseDir, filename);
        if (filePath == null) {
            sendError(sock, "INVALID_PATH");
            return;
        }
        Path parent = filePath.getParent();
        if (parent != null) {
            Files.createDirectories(parent);
        }

        try (RandomAccessFile raf = new RandomAccessFile(filePath.toFile(), "rw")) {
            raf.setLength(fileSize);

            for (int i = 0; i < blocksCount; i++) {
                byte[] frame = Protocol.recvMessage(sock);
                if (frame.length == 0)  { sendError(sock, "EMPTY_BLOCK");           return; }
                if (frame.length < 8)   { sendError(sock, "INVALID_BLOCK_HEADER");  return; }

                ByteBuffer bb = ByteBuffer.wrap(frame).order(ByteOrder.BIG_ENDIAN);
                int index   = bb.getInt();
                int dataLen = bb.getInt();
                int compLen = frame.length - 8;
                if (dataLen != compLen) { sendError(sock, "INVALID_BLOCK_LENGTH");  return; }

                byte[] compressed = new byte[compLen];
                bb.get(compressed);

                byte[] data;
                try {
                    data = Zlib.decompress(compressed);
                } catch (IOException e) {
                    sendError(sock, "DECOMPRESSION_FAILED");
                    return;
                }
                if (data.length > blockSize) {
                    sendError(sock, "INVALID_BLOCK_SIZE");
                    return;
                }
                if ((long) index * blockSize + data.length > fileSize) {
                    sendError(sock, "INVALID_BLOCK_RANGE");
                    return;
                }

                raf.seek((long) index * blockSize);
                raf.write(data);
            }
        }

        String computedHash = BlockAnalysis.computeFileHash(filePath, blockSize);
        if (!CryptoUtil.constantTimeEquals(computedHash, fullHash)) {
            sendError(sock, "HASH_MISMATCH");
            return;
        }

        JSONObject ok = new JSONObject().put("status", "OK").put("applied_blocks", blocksCount);
        Protocol.sendMessage(sock, ok.toString().getBytes(StandardCharsets.UTF_8));
    }

    /**
     * Resolve {@code filename} relative to {@code baseDir} and reject any
     * path that escapes the base directory.
     */
    private static Path resolveFilePath(Path baseDir, String filename) {
        Path resolved = baseDir.resolve(filename).toAbsolutePath().normalize();
        if (!resolved.startsWith(baseDir)) {
            return null;
        }
        return resolved;
    }

    private static void sendError(Socket sock, String code) throws IOException {
        Protocol.sendMessage(sock,
                new JSONObject().put("error", code).toString().getBytes(StandardCharsets.UTF_8));
    }
}
